extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,363 @@
+ * limitations under the License.
+ */
+
+import {AmpEvents} from '../../../src/amp-events';
+import {CSS} from '../../../build/amp-mega-menu-0.1.css';
+import {Keys} from '../../../src/utils/key-codes';
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {
+  closestAncestorElementBySelector,
+  isRTL,
+  scopedQuerySelector,
+  scopedQuerySelectorAll,
+  tryFocus,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {setModalAsClosed, setModalAsOpen} from '../../../src/modal';
+import {toArray} from '../../../src/types';
+
+/** @const {string} */
+const TAG = 'amp-mega-menu';
+
+/** @const {string} */
+const SCREEN_READER_CLOSE_LABEL = 'Close the mega menu';
+
+/**
+ * A mega menu component suitable for displaying large collections of links on
+ * desktop/tablets. An amp-list can be nested inside this component to enable
+ * template rendering via dynamically fetched data.
+ */
+export class AmpMegaMenu extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Array<!Element>} */
+    this.items_ = null;
+
+    /** @private {number} */
+    this.itemCount_ = 0;
+
+    /** @private {?Element} */
+    this.expandedItem_ = null;
+
+    /** @private {?Element} */
+    this.maskElement_ = null;
+
+    /** @private @const {!Document} */
+    this.document_ = this.win.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id || Math.floor(Math.random() * 100);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout === Layout.FIXED_HEIGHT;
+  }
+
+  /** @override */
+  buildCallback() {
+    // TODO(#24898): remove this assert when cleaning up experiment post launch.
+    userAssert(
+      isExperimentOn(this.win, 'amp-mega-menu'),
+      'Turning on the amp-mega-menu experiment is necessary to use the amp-mega-menu component.'
+    );
+
+    this.action_ = Services.actionServiceForDoc(this.element);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.registerMenuItems_();
+    // items may not be present after build if dynamically rendered via amp-list,
+    // in which case register them after DOM update instead.
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
+      this.registerMenuItems_();
+    });
+
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
```

Comment:
For reference, `amp-sidebar` appends it to the document's body

Reply:
Without header the mask will be appended to the mega menu element itself; this would still keep all elements inside the mega menu above the mask (though user might need to set background color on the menu items for it to look right). I believe with #24827 we want to move away from appending things directly under body, due to concerns from amp-story (even though in this case mega menu will probably not be used within stories).

Reply:
Could you please test and see if the behavior works correctly without a header element? It seems to me like it wouldn't given that the mega menu is position relative and the mask inside it will be absolute but relative to it

Reply:
I've made a [quick recording](https://vimeo.com/364689577) of what it looks like without a header. The mask has position fixed, so it should still cover the entire page (if that's the issue).